chain/chain/src/flat_storage_creator.rs

```diff
@@ -83,19 +171,170 @@ impl FlatStorageShardCreator {
 
                     // We continue saving deltas, and also start fetching state.
                     let block_hash = final_head.last_block_hash;
+                    let store = self.runtime_adapter.store().clone();
+                    let epoch_id = self.runtime_adapter.get_epoch_id(&block_hash)?;
+                    let shard_uid = self.runtime_adapter.shard_id_to_uid(shard_id, &epoch_id)?;
+                    let trie_storage = TrieDBStorage::new(store.clone(), shard_uid);
+                    let state_root =
+                        chain_store.get_chunk_extra(&block_hash, &shard_uid)?.state_root().clone();
+                    let trie = Trie::new(Box::new(trie_storage), state_root, None);
+                    let root_node = trie.retrieve_root_node().unwrap();
+                    let num_state_parts =
+                        root_node.memory_usage / STATE_PART_MEMORY_LIMIT.as_u64() + 1;
+                    let status = FetchingStateStatus {
+                        part_id: 0,
+                        num_parts_in_step: NUM_PARTS_IN_ONE_STEP,
+                        num_parts: num_state_parts,
+                    };
+                    debug!(target: "store", ?status, "Switching status to fetching state");
+
                     let mut store_update = chain_store.store().store_update();
                     store_helper::set_flat_head(&mut store_update, shard_id, &block_hash);
-                    store_helper::set_fetching_state_step(&mut store_update, shard_id, 0u64);
+                    store_helper::set_fetching_state_status(&mut store_update, shard_id, status);
```

Comment:
Should we store flat head here with a different prefix, maybe something like `set_temporary_flat_head` to distinguish it from when flat head is actually set and the flat state is ready to use? This way, there is no way the code can accidentally create a flat storage if flat state is not ready.

Reply:
We probably should. Let's do it separately because this change is already quite big and write some unit test checking that FS can't be created accidentally.